src/Mod/Part/App/PropertyTopoShapeList.h

```diff
@@ -0,0 +1,99 @@
+ *                                                                         *
+ *   This file is part of the FreeCAD CAx development system.              *
+ *                                                                         *
+ *   This library is free software; you can redistribute it and/or         *
+ *   modify it under the terms of the GNU Library General Public           *
+ *   License as published by the Free Software Foundation; either          *
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef PropertyTopoShapeList_H
+#define PropertyTopoShapeList_H
+
+#include <Mod/Part/PartGlobal.h>
+
+#include <vector>
+
+#include <App/Property.h>
+
+
+#include  "TopoShape.h"
+
+
+namespace Base {
+class Writer;
+class XMLReader;
+}
+
+namespace Part
+{
+class TopoShape;
+
+class PartExport PropertyTopoShapeList: public App::PropertyLists
+{
+    TYPESYSTEM_HEADER_WITH_OVERRIDE();
+
+public:
+    /**
+     * A constructor.
+     * A more elaborate description of the constructor.
+     */
+    PropertyTopoShapeList();
+
+    /**
+     * A destructor.
+     * A more elaborate description of the destructor.
+     */
+    ~PropertyTopoShapeList() override;
+
+    void setSize(int newSize) override;
+    int getSize() const override;
+
+    /** Sets the property
+     */
+    void setValue(TopoShape*);
+    void setValues(const std::vector<TopoShape*>&);
+
+    /// index operator
+    const TopoShape* operator[] (const int idx) const {
+        return _lValueList[idx];
+    }
+
+    const std::vector<TopoShape*>& getValues() const {
+        return _lValueList;
+    }
+
+    PyObject *getPyObject() override;
+    void setPyObject(PyObject *) override;
+
+    void Save(Base::Writer &writer) const override;
+    void Restore(Base::XMLReader &reader) override;
+
+    App::Property *Copy() const override;
+    void Paste(const App::Property &from) override;
+
+    unsigned int getMemSize() const override;
+
+private:
+    std::vector<TopoShape*> _lValueList;
```

Comment:
Is this on purpose to use a pointer here? If not then don't do it. Since the ownership of a TopoShape allocated on the heap is not always clear you cannot guarantee that PropertyTopoShapeList is the only owner of these objects. The result will be that several PropertyTopoShapeList own the same TopoShape instances which sooner or later will lead to a crash.

If you wanted to guarantee solely ownership you must change all methods to create a copy on the heap and them into the std::vector but then you can also just a TopoShape directly.

Don't worry about creating copies of a TopoShape because they are cheap. The class members of a TopoShape are a TopoDS_Shape, a long and an atomic int. The TopoDS_Shape itself is only a handle around the actual TShapes. So writing this won't create deep copies of the shape data:
```
TopoDS_Shape shape1 = ...
TopoDS_Shape shape2 = shape1;
TopoDS_Shape shape3 = shape2;
```

Reply:
Changed to const TopoShape& most places.   Could not get Restore to work without using pointers.  See PropertyTopoShapeList::Restore and PropertyTopoShapeList::afterRestore.

Reply:
OK I will have a deeper look on that...

Reply:
Thanks.  No big rush. The new code that depends on this isn't quite ready.

Reply:
I had a closer look at your implementation and it's fine. However, you should replace the raw pointer with shared pointers to be sure that no memory leak occurs.
There are a few places where `m_restorePointers.clear()` is called without actually deleting the pointers. So, with shared pointers we are safe.

What I have found are two missing aboutToSetValue() / hasSetValue() and for the Restore() function we can use TopoShape::Restore()

Here is the diff file of my modifications:
```
diff --git a/src/Mod/Part/App/PropertyTopoShapeList.cpp b/src/Mod/Part/App/PropertyTopoShapeList.cpp
index 9f4da48bdc..02791630ce 100644
--- a/src/Mod/Part/App/PropertyTopoShapeList.cpp
+++ b/src/Mod/Part/App/PropertyTopoShapeList.cpp
@@ -98,18 +98,22 @@ void PropertyTopoShapeList::setValues(const std::vector<TopoShape>& lValue)
 // clean out the list
 void PropertyTopoShapeList::clear()
 {
+    aboutToSetValue();
     _lValueList.clear();
     _lValueList.resize(0);
+    hasSetValue();
 }
 
 // populate the lists with the TopoShapes that have now finished restoring
 void PropertyTopoShapeList::afterRestore()
 {
+    aboutToSetValue();
     _lValueList.clear();
     for (auto entry : m_restorePointers) {
         _lValueList.push_back(*entry);
-        delete entry;     //??
     }
+    hasSetValue();
+
     m_restorePointers.clear();
     App::PropertyLists::afterRestore();
 }
@@ -170,14 +174,9 @@ void PropertyTopoShapeList::Restore(Base::XMLReader &reader)
     m_restorePointers.clear();      // just in case
     m_restorePointers.reserve(count);
     for (int i = 0; i < count; i++) {
-        reader.readElement("TopoShape");
-        std::string file (reader.getAttribute("file") );
-        if (!file.empty()) {
-            // add file to reader's list of files to be loaded by each newShape's RestoreDocFile()
-            TopoShape* newShape = new TopoShape();
-            m_restorePointers.push_back(newShape);
-            reader.addFile(file.c_str(), newShape);
-        }
+        auto newShape = std::make_shared<TopoShape>();
+        newShape->Restore(reader);
+        m_restorePointers.push_back(newShape);
     }
     reader.readEndElement("ShapeList");
 }
diff --git a/src/Mod/Part/App/PropertyTopoShapeList.h b/src/Mod/Part/App/PropertyTopoShapeList.h
index e1bbb388c0..add43270e9 100644
--- a/src/Mod/Part/App/PropertyTopoShapeList.h
+++ b/src/Mod/Part/App/PropertyTopoShapeList.h
@@ -97,7 +97,8 @@ private:
     // holds the new TopoShapes between their creation in Restore, and the completion of all
     // individual RestoreDocFile calls.  Once the restore is complete, the new TopoShapes are
     // inserted into _lValueList;
-    std::vector<TopoShape*> m_restorePointers;
+    using TopoShapePtr = std::shared_ptr<TopoShape>;
+    std::vector<TopoShapePtr> m_restorePointers;
 };
 
 } // namespace Part

```

What would be good too is to add a unit test to see that the property works. Here is some unpolished code that must be brought in form. If you don't know how to achieve that then ping me and I will write a proper test after this PR is merged.
```
doc = App.newDocument()
obj = doc.addObject("App::FeaturePython", "Object")
obj.addProperty("Part::PropertyTopoShapeList", "Shapes")
box = Part.makeBox(1,1,1)
box2 = Part.makeBox(1,1,2)
box3 = Part.makeBox(1,1,3)
obj.Shapes = [box, box2, box3]

doc.saveAs("/tmp/test.FCStd")
App.closeDocument(doc.Name)

doc = App.openDocument("/tmp/test.FCStd")
boxes = doc.ActiveObject.Shapes
boxes[0].Volume # ~1
boxes[1].Volume # ~2
boxes[2].Volume # ~3

twoboxes = [boxes[1], boxes[2]]

doc.openTransaction("Change shapes")
doc.ActiveObject.Shapes = twoboxes
len(doc.ActiveObject.Shapes) == 2
doc.commitTransaction()

doc.undo()

len(doc.ActiveObject.Shapes) == 3
```